youtube_dl/extractor/yospace.py

```diff
@@ -0,0 +1,202 @@
+from __future__ import unicode_literals
+
+import re
+
+from .common import InfoExtractor
+from ..utils import (
+    ExtractorError,
+    js_to_json,
+    float_or_none,
+    int_or_none,
+)
+
+
+class YospaceIE(InfoExtractor):
+    _VALID_URL = r'http://(?:csm-[a-z]|mas-[a-z]).cds\d+.yospace.com/(?P<type>csm|mas)/(?P<id>\d+/\d+)'
+    _TESTS = [
+        {
+            'url': 'http://csm-e.cds1.yospace.com/csm/108986746/100312513735',
+            'info_dict': {
+                'id': '108986746_100312513735',
+                'ext': 'mp4',
+                'title': '108986746_100312513735',
+                'description': None,
+            },
+            'params': {
+                # m3u8 download
+                'skip_download': True,
+            },
+        },
+    ]
+
+    def _extract_m3u8(self, hls_url):
+        formats = self._extract_m3u8_formats(hls_url, 'hls', 'mp4', m3u8_id='hls')
+        return formats
+
+    def _extract_formats(self, mas_url, video_id):
+        formats = []
+        hls_url = None
+        jfpage = self._download_webpage(mas_url, video_id)
+        jf = self._parse_json(jfpage, video_id, transform_source=js_to_json)
+        for ent in jf:
+            if ent.get('type', '') == 'application/x-mpeg-url':
+                hls_url = ent.get('url')
+                formats.extend(self._extract_m3u8(hls_url))
+            else:
+                tbr = float_or_none(ent.get('size', 0), 1000)
+                if tbr == 0:
+                    r = re.search(r'[\?\&]q=(\d+)', ent.get('url'))
+                    if r:
+                        tbr = float_or_none(r.group(1), 1)
+                formats.append({
+                    'url': ent.get('url'),
+                    'format_id': ent.get('method', 'unknown') + '-' + ent.get('container', 'unknown'),
+                    'protocol': ent.get('url').split(':')[0],
+                    'tbr': tbr,
+                    'ext': ent.get('container', 'unknown')
+                })
+        return formats
+
+    def _real_extract(self, url):
+        mobj = re.match(self._VALID_URL, url)
+        video_id = mobj.group('id')
+        url_type = mobj.group('type')
+        display_id = url_type
+        formats = []
+        hls_url = None
+
+        if url_type == 'mas':
+            mas_url = url.split('?')[0] + '?trans=json'
+            formats = self._extract_formats(mas_url, video_id)
+        else:
+            hls_url = url
+
+        if hls_url is not None:
+            formats.extend(self._extract_m3u8(hls_url))
+
+        self._sort_formats(formats)
+
+        return {
+            'id': video_id.replace('/', '_'),
+            'display_id': display_id,
+            'title': video_id.replace('/', '_'),
```

Comment:
Since this extractor by itself can't provide too much info, maybe it would be better to remove the _real_extract and _VALID_URL, don't import it in `__init__` and do something similar to `MTVServicesInfoExtractor`, which is the base class for the extractors that need it.